internal/controller/resource_manager.go

```diff
@@ -144,74 +147,124 @@ func getErrFromWatchEvent(event watch.Event) error {
 // watchEvents watches Kubernetes resources. The second part of the of the "list
 // and watch" strategy.
 // Any occurring errors are sent through errChanWatch. The watch stops when ctx is cancelled.
-func (r *ResourceManager) watchEvents(ctx context.Context, apiGroup string, watchOptions metav1.ListOptions, mutationChan chan *pb.KubernetesResourceMutation) error {
-	logger := r.logger.With(zap.String("api_group", apiGroup))
+func (r *ResourceManager) watchEvents(ctx context.Context, resourceVersion string, mutationChan chan *pb.KubernetesResourceMutation) error {
+	logger := r.logger
 
-	objGVR := schema.GroupVersionResource{Group: apiGroup, Version: "v1", Resource: r.resourceName}
+	// Build watch options from the provided resource version
+	watchOptions := metav1.ListOptions{
+		Watch:               true,
+		ResourceVersion:     resourceVersion,
+		AllowWatchBookmarks: true,
+	}
 
-	watcher, err := r.dynamicClient.Resource(objGVR).Namespace(metav1.NamespaceAll).Watch(ctx, watchOptions)
-	if err != nil {
-		logger.Error("Error setting up watch on resource", zap.Error(err))
+	lastKnownResourceVersion := resourceVersion
 
-		return err
-	}
+	var watcher watch.Interface
+
+	defer func() {
+		if watcher != nil {
+			watcher.Stop()
+		}
+	}()
 
 	mutationCount := 0
 
+	for {
+		// (Re)initialize the watcher using a helper
+		w, err := r.startWatcher(ctx, r.apiGroup, lastKnownResourceVersion, watchOptions, watcher, logger)
+		if err != nil {
+			return err
+		}
+
+		watcher = w
+
+		// Run the watcher loop until a restart is requested or an error occurs
+		updatedRV, restart, err := r.runWatcherUntilRestart(ctx, watcher, lastKnownResourceVersion, mutationChan, logger, &mutationCount)
+		if err != nil {
+			return err
+		}
+
+		lastKnownResourceVersion = updatedRV
+
+		if !restart {
+			return nil
+		}
+	}
+}
+
+// runWatcherUntilRestart processes events from a watcher until it needs to be restarted or an error occurs.
+// It returns the latest resourceVersion, whether a restart is needed, and any error encountered.
+func (r *ResourceManager) runWatcherUntilRestart(
+	ctx context.Context,
+	watcher watch.Interface,
+	lastKnownResourceVersion string,
+	mutationChan chan *pb.KubernetesResourceMutation,
+	logger *zap.Logger,
+	mutationCount *int,
+) (string, bool, error) {
```

Comment:
Good idea of creating this function. But you can simplify even further.
I would move the implementation of `startWatcher` here, which would allow calling `watcher.Stop()` in a `defer` here, which would be very clean.

There's actually very little code to move.

```suggestion
func (r *ResourceManager) runWatcherUntilRestart(
	ctx context.Context,
	lastKnownResourceVersion string,
	mutationChan chan *pb.KubernetesResourceMutation,
	mutationCount *int,
) (string, bool, error) {
	watchOptions := metav1.ListOptions{
		Watch:               true,
		ResourceVersion:     lastKnownResourceVersion,
		AllowWatchBookmarks: true,
	}

	objGVR := schema.GroupVersionResource{Group: apiGroup, Version: "v1", Resource: r.resourceName}

	watcher, err := r.dynamicClient.Resource(objGVR).Namespace(metav1.NamespaceAll).Watch(ctx, watchOptions)
	if err != nil {
		logger.Error("Error setting up watch on resource", zap.Error(err))

		return nil, err
	}

	defer watcher.Stop()
```